### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by adambrown1**
Introduce new plugins for

- Uni Freiburg weather service (pulled from [https://weather.uni-freiburg.de](https://weather.uni-freiburg.de))
- Teledyne flow controller (allowing reading and control)
- Scales with direct UDP communication (no separate display unit) - this was a pain!

Also improvements for WUTs

### CODE REVIEW

File: lanscale.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    commands = {
+        'tare': '0001E01020000',
+        'untare': '0001E11030000',
+        'zero': '0001E10300000'
+    }
+
+    def process_one_value(self, name=None, data=None):
+        """
+        Takes the raw data as returned by send_recv and parses
+        it for the float. Only for the scales.
```
Comment (jarongrigat) [inline lanscale.py]: What does the 'Only for the scales' mean?

File: lanscale.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        """
+        return float(re.search('(?P<value>\-?[0-9]+(?:\.[0-9]+)?)kg', data).group('value'))
+
+
+    def setup(self):
+        self.packet_bytes = 1024
+        self._device = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.logger.debug('Got somewhere')
```
Comment (jarongrigat) [inline lanscale.py]: remove these debug msgs?

File: lanscale.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        self._msg_start = '\u0002'
+        self._msg_end = '\u0003\r\n'
+        return True
+
+    def shutdown(self):
+        pass
+
+    def execute_command(self, quantity, value):
+        return quantity
+        if quantity.startswith('0001R'):
```
Comment (jarongrigat) [inline lanscale.py:43]: this code is never reached

File: teledyne_hfc.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    eol = b'\r>'
+    _msg_end = '\r'
+    commands = {  # these are not case sensitive
+        'set_setpoint': 'V4={value}',
+        'set_valvemode': 'V1={value}'
+    }
+    value_pattern = re.compile(f'(?P<value>{utils.number_regex})'.encode())
+
+    def execute_command(self, quantity, value):
+        self.logger.debug(f'Q {quantity} and v {value}')
```
Comment (jarongrigat) [inline teledyne_hfc.py]: consider removing

File: weather.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        auth = command.get('auth')
+        if isinstance(auth, list):
+            auth = tuple(auth)
+        if command['type'] == 'post':
+            requests_function = requests.post
+        elif command['type'] == 'get':
+            requests_function = requests.get
+        else:
+             raise ValueError(f"Unsupported request type {command['type']}")
+        self.logger.debug(f"Url: {url}")
```
Comment (jarongrigat) [inline weather.py]: remove debug msgs

**Review by jarongrigat — APPROVED** [review]
Nice! I would just avoid the debug messages that get generated with every readout.

**Comment by jarongrigat** [comment]
Also, looking what CodeFactor says: I guess this `.bind('0.0.0.0', port)` and then checking whether it's the right device later on is unavoidable?

File: lanscale.py (hunk 2 above)
Comment (adambrown1) [reply lanscale.py]: Removed in new version

File: lanscale.py (hunk 1 above)
Comment (adambrown1) [reply lanscale.py]: I have no idea. It is gone

File: weather.py (hunk 5 above)
Comment (adambrown1) [reply weather.py]: Done

**Comment by adambrown1** [comment]
> Also, looking what CodeFactor says: I guess this `.bind('0.0.0.0', port)` and then checking whether it's the right device later on is unavoidable?

I spent ages trying to understand this and still don't really. I think the IP address is your local IP address, and you could in theory specify the local IP address to bind to (in case your machine is connected via multiple network interfaces). But that's not what we want here.

The reason it is tricky is because they all send data back to the same port. We could also configure them not to do that, but I think that gets even more confusing (we still need a custom plugin anyway) and you use up more ports (although we're unlikely to ever have so many scales that's an issue!)

In any case I have had to almost completely rewrite the plugin because I realised messages were building up in the queue overnight - sorry. I hope now it works but maybe you want to wait a few days before reviewing it!

File: teledyne_hfc.py
Comment (TiffanyLuce) [inline teledyne_hfc.py:1]: You forgot to change the description of this guy from "Scale for nitrogen dewar in pancake."

Comment (adambrown1) [reply teledyne_hfc.py:1]: Whoopsie!

**Review by jarongrigat — APPROVED** [review]

**Pull request merged**